### PR TITLE
vulkan-loader 1.3.220 (new formula)

### PR DESCRIPTION
**Pull request by rcombs**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Replaces #39527.

Also includes a fix to the `molten-vk` formula to make it compatible with the loader.

### CODE REVIEW

File: Formula/molten-vk.rb
Comment (SMillerDev) [inline Formula/molten-vk.rb:146]: Is this issue submitted upstream? What is it actually fixing?

File: Formula/vulkan-loader.rb
Comment (SMillerDev) [inline Formula/vulkan-loader.rb:10]: Can it use 3.10?

Comment (rcombs) [reply Formula/vulkan-loader.rb]: Sure, done.

File: Formula/molten-vk.rb
Comment (rcombs) [reply Formula/molten-vk.rb:146]: I'm not sure if an upstream submission would really be appropriate, as this is specific to installing system-wide, which the upstream package doesn't provide any support for (we do it all manually in the formula).

The issue is that the actual driver lib is installed in `{formula_prefix}/lib/libMoltenVK.dylib` (fairly sensibly), while the `icd.json` is installed in `{formula_prefix}/share/vulkan/icd.d/MoltenVK_icd.json` (also fairly sensible), and the `library_path` property needs to point to the driver lib relative to the dir containing the json. The upstream package just gives `./libMoltenVK.dylib`, which is appropriate when packaging MoltenVK in a .app bundle with the dylib and json files right next to each other (which is the main use-case they target).

Comment (carlocab) [reply Formula/molten-vk.rb:146]: Can we use `Pathname#relative_path_from` to make this understandable without having to check the directory structure manually (`grep relative_path_from` for examples)?

Comment (rcombs) [reply Formula/molten-vk.rb]: It can indeed! Pushed.

File: Formula/molten-vk.rb
Comment (carlocab) [inline Formula/molten-vk.rb:146]:
```suggestion
    inreplace "MoltenVK/icd/MoltenVK_icd.json",
              "./libMoltenVK.dylib",
              (lib/"libMoltenVK.dylib").relative_path_from(share/"vulkan/icd.d")
```

File: Formula/vulkan-loader.rb
Comment (carlocab) [inline Formula/vulkan-loader.rb:17]: Let's follow the pattern used by `brew create --cmake`.
```suggestion
    system "cmake", "-S", ".", "-B", "build",
                    "-DVULKAN_HEADERS_INSTALL_DIR=#{Formula["vulkan-headers"].opt_prefix}",
                    "-DFALLBACK_DATA_DIRS=#{HOMEBREW_PREFIX}/share:/usr/local/share:/usr/share",
                    *std_cmake_args
    system "cmake", "--build", "build"
    system "cmake", "--build", "install"
```

File: Formula/vulkan-loader.rb
Comment (carlocab) [inline Formula/vulkan-loader.rb:15]: What is `FALLBACK_DATA_DIRS` used for?

File: Formula/vulkan-loader.rb
Comment (carlocab) [inline Formula/vulkan-loader.rb:19]: Why do we need to put `Cflags: ` in a capture group? If the intent is to replace only that substring, then I suspect this doesn't fulfil that.

**Review by carlocab — COMMENTED** [review]
Bump the `revision` for `molten-vk`, please.

File: Formula/vulkan-loader.rb
Comment (carlocab) [inline Formula/vulkan-loader.rb:10]:
```suggestion
  depends_on "cmake" => :build
  depends_on "pkg-config" => :build
  depends_on "python@3.10" => :build
```

File: Formula/molten-vk.rb
Comment (carlocab) [inline Formula/molten-vk.rb:149]: I just noticed that this change seems to be completely independent of `vulkan-loader`. Is that correct? If so, please move this change into a separate PR.

Comment (rcombs) [reply Formula/molten-vk.rb:149]: Neither _technically_ depends on the other (e.g. on linux, you'd want `vulkan-loader` but can't have `molten-vk`, and on macOS, it's entirely legal to just link directly against `libMoltenVK.dylib` and skip `vulkan-loader` altogether). But `vulkan-loader`'s purpose is to dynamically load and dispatch to Vulkan ICDs at runtime, and `molten-vk` is the Vulkan ICD for macOS, so the `vulkan-loader` package on macOS is useless (and can't have its functionality properly tested) without a working `molten-vk` ICD registry JSON file. Meanwhile, on Linux, you'd use it to load a different driver (e.g. mesa). On the other hand, the ICD registry JSON file is only useful when using `vulkan-loader`.

I can move this to a separate PR if you'd prefer, but they're in a sort of "neither can be tested without the other" situation, so I figured it made sense to bundle them.

Comment (carlocab) [reply Formula/molten-vk.rb:149]: Do you need `molten-vk` to be installed when testing `vulkan-loader, at least on macOS? Because it won't be installed unless you declare a dependency.

Comment (rcombs) [reply Formula/molten-vk.rb:149]: `brew test vulkan-loader` will work fine without it, but more complex functional testing (i.e. something that uses Vulkan to draw to a window) wouldn't. So you can test that the package is linked properly, but not actually _use_ it for anything. (And I don't think it'd be appropriate to have the test program try to draw something, since (b) it'd be absurdly complex, and (b) it wouldn't work anywhere where a suitable driver isn't necessarily available, e.g. potentially linux).

File: Formula/vulkan-loader.rb
Comment (carlocab) [reply Formula/vulkan-loader.rb]: Shouldn't we have
```ruby
Formula["vulkan-headers"].opt_share
```
in `FALLBACK_DATA_DIRS` then? Or is that made redundant by `VULKAN_HEADERS_INSTALL_DIR`?

Comment (rcombs) [reply Formula/vulkan-loader.rb]: We could, but it'd be pretty unusual (and obviously inappropriate on Linux). Generally, the loader is expected to look in a list of standard dirs provided by the OS (or package manager) for _all_ drivers, and each driver is expected to install a .json file in one of those dirs; the loader isn't meant to have to know anything about each individual driver at compile-time. This mechanism also allows other packages to link against `vulkan-loader` without creating a heavier hard dependency on a driver, and lets the user enable or disable a particular driver at runtime by just linking or unlinking it.

The fact that there's only one Vulkan driver in Brew right now is sorta just a coincidence; Linuxbrew could expose several more from the `mesa` formula, there are at least two others for AMD GPUs (plus one for Nvidia), and it's entirely plausible that we could one day see multiple implementations on macOS (like how ANGLE and MoltenGL coexist).

We see similar behavior in the formulae for `pkg-config`, `automake`, `fish`, `zsh`, and a number of others that want to allow other packages to provide resources for them to use at runtime.

**Comment by rcombs** [comment]
Anything in particular this is waiting on at this point? FYI, after this is landed I'm planning on also adding formulae for shaderc and libplacebo, and adding HEAD support for molten-vk.

**Comment by rcombs** [comment]
(Rebased without changes)

**Comment by SMillerDev** [comment]
There's still unresolved comments, some of them waiting for action on your side (like the capture group)

File: Formula/vulkan-loader.rb
Comment (rcombs) [reply Formula/vulkan-loader.rb]: Removed.

**Comment by rcombs** [comment]
Anything else still at me?

**Review by carlocab — APPROVED** [review]
Thanks! A few more questions; we can fix those in a follow up if needed.

File: Formula/vulkan-loader.rb
Comment (rcombs) [reply Formula/vulkan-loader.rb:24]: Hmmmm, it looks like in practice the strings we end up with by default are just `/etc` and `/etc/xdg`, which aren't too unreasonable; I could see maybe setting `CMAKE_INSTALL_FULL_SYSCONFDIR` to `#{HOMEBREW_PREFIX}/etc` (which would leave `/etc` still being searched as well), and `FALLBACK_CONFIG_DIRS` to `#{HOMEBREW_PREFIX}/etc/xdg:/etc/xdg`, but… is anyone actually going to stick local config files there (vs in `/etc[/xdg]`) anyway?

File: Formula/vulkan-loader.rb
Comment (rcombs) [reply Formula/vulkan-loader.rb:29]: Seems reasonable, though currently that just gets you a `-I`, which at best is useless and at worst would generate a compiler warning for the nonexistent directory. Maybe just hold off and consider doing something here if the package ever adds something? (I don't really expect it tbh, since this package really just provides the runtime lib rather than headers.)

File: Formula/vulkan-loader.rb
Comment (carlocab) [reply Formula/vulkan-loader.rb:24]: Actually I think it looks like in practice we get a path like
```
$HOMEBREW_PREFIX/Cellar/vulkan-loader/x.y.z/etc
```
which isn't great for configuration files because it gets deleted by `brew upgrade vulkan-loader`.

> but… is anyone actually going to stick local config files there (vs in `/etc[/xdg]`) anyway?

Writing to `/etc` requires `sudo`, and we aim to minimise needing it when using Homebrew.

Comment (rcombs) [reply Formula/vulkan-loader.rb:24]: Oh, hah, it's even sillier than that, actually: we get `/etc/opt/homebrew/Cellar/vulkan-loader/1.3.220`. Yeah, that's pretty obviously nonsensical, I'll try to write a patch for it.

Comment (carlocab) [reply Formula/vulkan-loader.rb:24]: Yea that's nonsense (and actually is inconsistent with [CMake documentation](https://cmake.org/cmake/help/v3.24/module/GNUInstallDirs.html#special-cases)).

It should be enough to set `SYSCONF` and/or `FALLBACK_CONFIG_DIRS`, I think.

**Pull request closed without merging**
